src/Silicon.cpp

```diff
@@ -874,124 +1135,493 @@ namespace galsim {
+        Bounds<double>* pixelOuterBoundsData = _pixelOuterBounds.data();
+        Position<float>* horizontalBoundaryPointsData = _horizontalBoundaryPoints.data();
+        Position<float>* verticalBoundaryPointsData = _verticalBoundaryPoints.data();
+
+        double* abs_length_table_data = _abs_length_table_GPU.data();
+        
+        Position<double>* emptypolyData = _emptypolyGPU.data();
+
 #ifdef _OPENMP
-#pragma omp parallel reduction (+:addedFlux)
-        {
-#pragma omp for
+#ifndef GALSIM_USE_GPU
+#pragma omp parallel for reduction(+:addedFlux)
+#else
+#pragma omp target teams distribute parallel for map(to: photonsX[i1:i2-i1], photonsY[i1:i2-i1], photonsDXDZ[i1:i2-i1], photonsDYDZ[i1:i2-i1], photonsFlux[i1:i2-i1], photonsWavelength[i1:i2-i1], diffStepRandomArray[0:(i2-i1)*2], conversionDepthRandomArray[0:i2-i1], pixelNotFoundRandomArray[0:i2-i1]) reduction(+:addedFlux)
 #endif
-        for (int i=i1; i<i2; i++) {
-            // Get the location where the photon strikes the silicon:
-            double x0 = photons.getX(i); // in pixels
-            double y0 = photons.getY(i); // in pixels
-            xdbg<<"x0,y0 = "<<x0<<','<<y0;
-
-            double dz = calculateConversionDepth(photons, i, conversionDepthRandom[i-i1]);
-            if (photons.hasAllocatedAngles()) {
-                double dxdz = photons.getDXDZ(i);
-                double dydz = photons.getDYDZ(i);
-                double dz_pixel = dz * invPixelSize;
-                x0 += dxdz * dz_pixel; // dx in pixels
-                y0 += dydz * dz_pixel; // dy in pixels
-            }
-            xdbg<<" => "<<x0<<','<<y0;
-            // This is the reverse of depth. zconv is how far above the substrate the e- converts.
-            double zconv = _sensorThickness - dz;
-            xdbg<<"zconv = "<<zconv<<std::endl;
-            if (zconv < 0.0) continue; // Throw photon away if it hits the bottom
-            // TODO: Do something more realistic if it hits the bottom.
-
-            // Now we add in a displacement due to diffusion
-            if (_diffStep != 0.) {
-                double diffStep = std::max(0.0, diffStep_pixel_z * std::sqrt(zconv * _sensorThickness));
-                x0 += diffStep * diffStepRandom[(i-i1)*2];
-                y0 += diffStep * diffStepRandom[(i-i1)*2+1];
+#endif
+	for (int i = i1; i < i2; i++) {
+	    double x0 = photonsX[i];
+	    double y0 = photonsY[i];
+
+	    // calculateConversionDepth
+	    double dz;
+	    if (photonsHasAllocatedWavelengths) {
+		double lambda = photonsWavelength[i];
+
+		// perform abs_length_table lookup with linear interpolation
+		int tableIdx = int((lambda - 255.0) / 5.0);
+		double alpha = (lambda - ((float(tableIdx) * 5.0) + 255.0)) / 5.0;
+                if (tableIdx < 0) tableIdx = 0;
+		int tableIdx1 = tableIdx + 1;
+                if (tableIdx > 239) tableIdx = 239;
+                if (tableIdx1 > 239) tableIdx1 = 239;
```

Comment:
Too many special numbers here.  255.0, 5.0, 239.  These all need to be computed from the input abs_length_table you get from python.  Don't assume these are necessarily the values forever.